### PR TITLE
Port raindrops test-suite to support Data.Text

### PR DESCRIPTION
**Pull request by tejasbubane**
Issue: https://github.com/exercism/haskell/issues/841

### CODE REVIEW

File: exercises/raindrops/test/Tests.hs [hunk 1]
```diff
@@ -1,10 +1,12 @@
 {-# OPTIONS_GHC -fno-warn-type-defaults #-}
+{-# LANGUAGE OverloadedStrings #-}
 
 import Data.Foldable     (for_)
 import Test.Hspec        (Spec, describe, it, shouldBe)
 import Test.Hspec.Runner (configFastFail, defaultConfig, hspecWith)
+import Data.String       (fromString)
 
-import Raindrops (convert)
+import Raindrops         (convert)
```
Comment (sshine) [inline exercises/raindrops/test/Tests.hs]: Even though this is not a consistent standard in this track, or any recommended practice outside of this track, I've mostly seen that the module being tested is listed outside of alphabetical order and without indentation. Whether we should keep this convention or not is clearly debatable, but it should be outside the scope of this PR.

```
$ ls exercises
114
$ find . -name Tests.hs -exec ack '^import (?!Test)\S+ \(' {} \; | wc -l
100
```

```suggestion
import Raindrops (convert)
```

File: exercises/raindrops/test/Tests.hs [hunk 2]
```diff
@@ -16,7 +18,7 @@ specs = describe "convert" $ for_ cases test
     test (number, expected) = it description assertion
       where
         description = show number
-        assertion   = convert number `shouldBe` expected
+        assertion   = convert number `shouldBe` (fromString expected)
```
Comment (sshine) [inline exercises/raindrops/test/Tests.hs]: CI throws up about this.

```suggestion
        assertion   = convert number `shouldBe` fromString expected
```

File: exercises/raindrops/examples/success-text/src/Raindrops.hs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import           Data.Maybe (fromMaybe)
+import qualified Data.Text as T
+import           Data.Text (Text)
+
+convert :: Int -> Text
+convert n = fromMaybe (T.pack $ show n) maybeSound
+  where
+    maybeSound = sound "Pling" 3 `mappend`
+                 sound "Plang" 5 `mappend`
+                 sound "Plong" 7
```
[the remark this answers is not in the document]
Comment (sshine) [reply exercises/raindrops/examples/success-text/src/Raindrops.hs]: Also, this is a very cool use of the `Semigroup a => Semigroup (Maybe a)` instance!

Comment (sshine) [inline exercises/raindrops/examples/success-text/src/Raindrops.hs]: You could also avoid the `maybeSound` binding by writing

```haskell
convert n = fromMaybe (T.pack (show n)) $
  sound "Pling" 3 `mappend`
  sound "Plang" 5 `mappend`
  sound "Plong" 7
  where
    sound = ...
```

File: exercises/raindrops/.meta/hints.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+- Read about [string types](https://haskell-lang.org/tutorial/string-types) in Haskell.
+- Add `- text` to your list of dependencies in package.yaml.
+- Import `Data.Text` in [the following way](https://hackernoon.com/4-steps-to-a-better-imports-list-in-haskell-43a3d868273c):
+
+```haskell
+import qualified Data.Text as T
+import           Data.Text (Text)
+```
+
+- You can now write e.g. `responseFor :: Text -> Text` and refer to `Data.Text` combinators as e.g. `T.isSuffixOf`.
```
[the remark this answers is not in the document]
Comment (sshine) [reply exercises/raindrops/.meta/hints.md]: Come to think of it, perhaps this is superfluous with the "You can then replace ..." sentence in a bullet below.

```suggestion
- You can now refer to `Data.Text` combinators as e.g. `T.pack`.
```

File: exercises/raindrops/examples/success-text/src/Raindrops.hs [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    sound noise factor | n `rem` factor == 0 = Just noise
+                       | otherwise           = Nothing
```
[the remark this answers is not in the document]
Comment (sshine) [reply exercises/raindrops/examples/success-text/src/Raindrops.hs]: After having reviewed #843, I've come to think that solutions like

```haskell
sound noise factor =
  if n `rem` factor == 0
  then pure noise
  else empty
```

are perfectly acceptable.

**Comment by tejasbubane** [comment]
@sshine Thanks 👍 I have made the changes.

**Comment by petertseng** [comment]
Well, "cannot" is a flexible term. There's always https://github.com/exercism/haskell/compare/b33ff3ea9c0b44398d9ea8c199af58e061bcbcd3..79a5a2b05383911d49680351a0a1212af98f8110.

**Comment by petertseng** [comment]
EBWODP: And may I point out that (on the web interface), the word "force-pushed" is a link to that.

**Comment by sshine** [comment]
Wow, I didn't know that was a hyperlink.

Thanks for that.

**Pull request merged**